Assets/Scripts/Effects/SpawnParticleEffect.cs

```diff
@@ -1,50 +1,44 @@
-﻿using System;
+﻿using MapTools;
 using UnityEngine;
 
 namespace Effects
 {
     public class SpawnParticleEffect : MonoBehaviour
     {
-        [SerializeField]
-        private GameObject spawnParticleEffect;
-        [SerializeField]
-        private bool onDestroy;
-        [SerializeField]
-        private bool onCollision;
-        [SerializeField]
-        private bool useTag;
-        [SerializeField]
-        private new string tag = string.Empty;
-        
-        
+        [SerializeField] private GameObject spawnParticleEffect;
+        [SerializeField] private bool onDestroy;
+        [SerializeField] private bool onCollision;
+        [SerializeField] private bool useTag;
+        [SerializeField] private new string tag = string.Empty;
+
         private void OnDestroy()
         {
             if (onDestroy)
-            { 
-                SpawnEffect();  
+            {
+                SpawnEffect();
             }
         }
 
         private void OnTriggerEnter2D(Collider2D col)
         {
             if (!onDestroy && onCollision && !useTag)
-            { 
-                SpawnEffect();  
+            {
+                SpawnEffect();
             }
-            if (!onDestroy && onCollision && useTag)
+
+            if (onDestroy || !onCollision || !useTag) return;
+            if (col.CompareTag(tag))
             {
-                if (col.CompareTag(tag))
-                {
-                    SpawnEffect();   
-                }
+                SpawnEffect();
             }
         }
 
         private void SpawnEffect()
         {
             if (!spawnParticleEffect.CompareTag("Circle"))
-            { 
-                SpawnObject.Spawn(spawnParticleEffect, transform.localPosition, transform.rotation);
+            {
+                var tempTransform = transform;
+                SpawnObject.Spawn(spawnParticleEffect, tempTransform.localPosition, tempTransform.rotation);
             }
             else
             {
```

Comment:
creates null pointer 
NullReferenceException: Object reference not set to an instance of an object
MapTools.SpawnObject.GetNewObject (UnityEngine.GameObject obj, UnityEngine.Vector3 location) (at Assets/Scripts/MapTools/SpawnObject.cs:37)
MapTools.SpawnObject.Spawn (UnityEngine.GameObject obj, UnityEngine.Vector3 location) (at Assets/Scripts/MapTools/SpawnObject.cs:26)
Effects.SpawnParticleEffect.SpawnEffect () (at Assets/Scripts/Effects/SpawnParticleEffect.cs:45)
Effects.SpawnParticleEffect.OnDestroy () (at Assets/Scripts/Effects/SpawnParticleEffect.cs:18)

Reply:
fixed